src/main/java/org/jenkinsci/plugins/workflow/graphanalysis/AbstractFlowScanner.java

```diff
@@ -0,0 +1,344 @@
+        }
+        return setup(Collections.singleton(head), blackList);
+    }
+
+    /**
+     * Helper: version of {@link #setup(Collection, Collection)} where we don't have any nodes to blacklist and have just a single head
+     */
+    public boolean setup(@CheckForNull FlowNode head) {
+        if (head == null) {
+            return false;
+        }
+        return setup(Collections.singleton(head), Collections.EMPTY_SET);
+    }
+
+    /** Reset internal state so that we can begin walking a new flow graph
+     *  Public APIs need to invoke this before searches */
+    protected abstract void reset();
+
+    /**
+     * Set up to begin flow scanning using the filteredHeads as starting points
+     *
+     * This method makes several assumptions:
+     *
+     *   - {@link #reset()} has already been invoked to reset state
+     *   - filteredHeads has already had any points in {@link #myBlackList} removed
+     *   - none of the filteredHeads are null
+     * @param filteredHeads Head nodes that have been filtered against blackList
+     */
+    protected abstract void setHeads(@Nonnull Collection<FlowNode> filteredHeads);
+
+    /**
+     * Actual meat of the iteration, get the next node to visit, using & updating state as needed
+     * @param current Current node to use in generating next value
+     * @param blackList Nodes that are not eligible for visiting
+     * @return Next node to visit, or null if we've exhausted the node list
+     */
+    @CheckForNull
+    protected abstract FlowNode next(@Nonnull FlowNode current, @Nonnull Collection<FlowNode> blackList);
+
+    @Override
+    public boolean hasNext() {
+        return myNext != null;
+    }
+
+    @Override
+    public FlowNode next() {
+        if (myNext == null) {
+            throw new NoSuchElementException();
+        }
+
+        myCurrent = myNext;
+        myNext = next(myCurrent, myBlackList);
+        return myCurrent;
+    }
+
+    @Override
+    public final void remove() {
+        throw new UnsupportedOperationException("FlowGraphs are immutable, so FlowScanners can't remove nodes");
+    }
+
+    @Override
+    @Nonnull
+    public Iterator<FlowNode> iterator() {
+        return this;
+    }
+
+    /**
+     * Expose a filtered view of this FlowScanner's output.
+     * @param filterCondition Filterator only returns {@link FlowNode}s matching this predicate.
+     * @return A {@link Filterator} against this FlowScanner, which can be filtered in additional ways.
+     */
+    @Override
+    @Nonnull
+    public Filterator<FlowNode> filter(@Nonnull Predicate<FlowNode> filterCondition) {
+        return new FilteratorImpl<FlowNode>(this, filterCondition);
+    }
+
+    /**
+     * Find the first FlowNode within the iteration order matching a given condition
+     * Includes null-checking on arguments to allow directly calling with unchecked inputs (simplifies use).
+     * @param heads Head nodes to start walking from
+     * @param endNodes
```

Comment:
:ant: Definitely this parameter should be documented. To start with, it is not at all obvious whether the bound is inclusive or exclusive (cf. `String.substring`). Then, what does it mean for this to be null? Same as an empty collection? And is there any required graph relationship between the start and end nodes, the way for example `substring` would throw `IndexOutOfBoundsException` if `beginIndex > endIndex`?